scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java

```diff
@@ -0,0 +1,110 @@
+package edu.asu.diging.scriptoocloud.core.service.impl;
+
+import edu.asu.diging.scriptoocloud.core.data.DataFileRepository;
+import edu.asu.diging.scriptoocloud.core.data.DatasetRepository;
+import edu.asu.diging.scriptoocloud.core.exceptions.FileSystemStorageException;
+import edu.asu.diging.scriptoocloud.core.model.impl.DataFile;
+import edu.asu.diging.scriptoocloud.core.model.impl.Dataset;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+/**
+ * Test class for DataFileService
+ *
+ * @author John Coronite
+ */
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = AnnotationConfigContextLoader.class)
+class DataFileServiceTest {
+
+    private final Long datasetId = 1L;
+    private final int pageSize = 25;
+    private final int pageNumber = 1;
+
+    @InjectMocks
+    private DataFileService dataFileService;
+
+    @Mock
+    private DataFileRepository dataFileRepository;
+
+    @Mock
+    private DatasetRepository datasetRepository;
+
+    @Mock
+    private FileSystemService fileSystemService;
+
+    @BeforeEach
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_createFile() throws FileSystemStorageException {
+
+        Dataset dataset = Mockito.mock(Dataset.class);
+        Mockito.when(datasetRepository.findById(datasetId))
+                .thenReturn(Optional.of(dataset));
+        Mockito.when(dataset.addFile(dataFileRepository.save(Mockito.any()))).thenReturn(1L);
```

Comment:
you only want 1L to be returned when a specific dataset is being saved, don't you? Not any?

Reply:
> you only want 1L to be returned when a specific dataset is being saved, don't you? Not any?

I'm having some general trouble with testing, and this test is a good example.
How would I best go about not using `Mockito.any()` above?
I looked into `ArgumentCaptor` and `ArgumentMatcher`, but I'm not sure if these could be of use.

I could return the `dataFile` created in the test method, and then use its Id:

```
Mockito.when(dataFileRepository.save(Mockito.any())).thenReturn(dataFile);
Mockito.when(dataset.addFile(dataFileRepository.save(dataFile))).thenReturn(dataFile.getId());
```

This works, but it still uses `Mockito.any()`, so it doesn't seem to be a viable solution.

Similarly, when running this test, `dataset.files` is null so both `dataset.add(file)` and `dataFileRepository.save(file)` throw an NPE.  
I tried reflection to add a Hibernate `PersistentSet` to the `dataset`, but received:
`org.hibernate.LazyInitializationException: failed to lazily initialize a collection, could not initialize proxy - no Session`
So, I then annotated the test class with `@Transactional`, and received the error:
`Failed to retrieve PlatformTransactionManager for @Transactional test...`
So, then I then added the `PersistenceConfig` class to the `@ContextConfiguration` annotation, but received an error that Spring couldn't create an `EntityManagerFactory`.

I added a `HashSet` instead of a Hibernate `PersistenceSet` for the files attribute on the `dataset` but this seems like a clunky solution.

Generally, these sorts of problems were why I was previously using mocks and spies, etc. but I could use some guidance as to what the best approach is for these sorts of tests.

Reply:
> How would I best go about not using Mockito.any() above?

Here is an example. 
https://github.com/diging/citesphere/blob/b0fec7c0eea580b3f8f1f8bfa09ef3d81b8008c2/citesphere/src/test/java/edu/asu/diging/citesphere/core/util/model/impl/CitationHelperTest.java#L205
I would create an argument matcher that is initialized for example with filename, type, and dataset id and checks these values agains the passed datafile object. If they match, it returns the id.

For the Nullpointer exception, this might actually a hint that there is a bug in your code or at least the potential for a bug. When is `files` initialized? I couldn't find it in either the dataset class, nor the datafile service or the dataset service. So, given that your test is throwing a NPE there could potentially a scenario that your code would throw that at runtime as well. I'm not sure, what Hibernate exactly does, but maybe it automatically adds a persistent set to any one to many relationship, but a) I wouldn't trust that this is always happening or that another framework would do the same (in case we ever exchange the ORM) and b) if at any point the code is changed in a way that a dataset does not have to be persisted yet to add a datafile, then this would also fail. 

Does that make sense?

Reply:
Thanks - I think that makes sense.  So, if I implement checking for a null file Set in the `Dataset` class (where the NPE was coming from) what implementation of `Set` should I use?  Presumably, Hibernate's `PersistentSet` is Hibernate-specific, but I see from the docs that the underlying collection is a `HashSet`.  Could I do:
```
@Override
    public Long addFile(DataFile dataFile) {
        if (files == null){
            files = new HashSet<>();
        }
        files.add(dataFile);
        return dataFile.getId();
    }
```

Reply:
Yes, I would just use a HashSet. Hibernate will replace that with a PersistentSet when you retrieve an object from the DB.